rewrite-java/src/main/java/org/openrewrite/java/RenamePackage.java

```diff
@@ -0,0 +1,313 @@
+            }
+            return i;
+        }
+
+        @Override
+        public J.MethodDeclaration visitMethodDeclaration(J.MethodDeclaration method, ExecutionContext ctx) {
+            J.MethodDeclaration m = super.visitMethodDeclaration(method, ctx);
+            m = m.withReturnTypeExpression(transformName(m.getReturnTypeExpression()));
+            return m.withThrows(m.getThrows() == null ? null : ListUtils.map(m.getThrows(), this::transformName));
+        }
+
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+            J.MethodInvocation m = super.visitMethodInvocation(method, ctx);
+
+            if (m.getSelect() instanceof NameTree
+                    && m.getType() != null && m.getType().hasFlags(Flag.Static)) {
+                m = m.withSelect(transformName(m.getSelect()));
+            }
+
+            if (m.getSelect() != null) {
+                JavaType.Class selectType = TypeUtils.asClass(m.getSelect().getType());
+                if (selectType != null && selectType.getFullyQualifiedName().equals(oldPackageName)) {
+                    m = m.withSelect(m.getSelect().withType(targetPackageType));
+                }
+            }
+
+            if (m.getType() != null) {
+                if (m.getType().getDeclaringType().getFullyQualifiedName().equals(oldPackageName)) {
+                    m = m.withDeclaringType(targetPackageType);
+                }
+            }
+
+            return m;
+        }
+
+        @Override
+        public J.MultiCatch visitMultiCatch(J.MultiCatch multiCatch, ExecutionContext ctx) {
+            final J.MultiCatch m = visitAndCast(multiCatch, ctx, super::visitMultiCatch);
+            return m.withAlternatives(ListUtils.map(m.getAlternatives(), this::transformName));
+        }
+
+        @Override
+        public J.NewArray visitNewArray(J.NewArray newArray, ExecutionContext ctx) {
+            final J.NewArray n = super.visitNewArray(newArray, ctx);
+            return n.withTypeExpression(transformName(n.getTypeExpression()));
+        }
+
+        @Override
+        public J.NewClass visitNewClass(J.NewClass newClass, ExecutionContext ctx) {
+            final J.NewClass n = super.visitNewClass(newClass, ctx);
+            return n.withClazz(transformName(n.getClazz()));
+        }
+
+        @Override
+        public J.Package visitPackage(J.Package pkg, ExecutionContext context) {
+            final String original = pkg.getExpression().printTrimmed().replaceAll("\\s", "");
+            getCursor().putMessageOnFirstEnclosing(J.CompilationUnit.class, CHANGING_FROM_KEY, original);
+
+            if (original.equals(oldPackageName)) {
+                getCursor().putMessageOnFirstEnclosing(J.CompilationUnit.class,CHANGING_TO_KEY, targetPackageType.getFullyQualifiedName());
+                pkg = pkg.withTemplate(newPackageExpr, pkg.getCoordinates().replace(), targetPackageType.getFullyQualifiedName());
```

Comment:
here you can reference the ThreadLocal JavaParser in the template
```suggestion
                return pkg.withTemplate(template("package #{}").javaParser(PARSER.get()).build(), pkg.getCoordinates().replace(), newFullyQualifiedPackageName);
```

Reply:
I added the JavaParser directly to the built template, since it's called in multiple places.